starknet-rpc-test/src/lib.rs

```diff
@@ -129,45 +134,60 @@ impl Drop for MadaraClient {
     }
 }
 
-fn get_free_port() -> Result<u16, TestError> {
-    for port in MIN_PORT..=MAX_PORT {
-        if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
-            return Ok(listener.local_addr().expect("No local addr").port());
+fn find_available_ports() -> Result<NodePorts, TestError> {
+    let mut available_ports = Vec::new();
+
+    for index in 0..3 {
+        let mut selected_port = 0;
+        let mut port = 1024 + index * 20000 + (std::process::id() % 20000) as u16;
+
+        while selected_port == 0 && port < ENDING_PORT {
+            if TcpListener::bind(("127.0.0.1", port)).is_ok() {
+                selected_port = port;
+            }
+            port += 1;
         }
-        // otherwise port is occupied
+
+        if selected_port == 0 {
+            return Err(TestError::NoFreePorts);
+        }
+
+        available_ports.push(selected_port);
     }
 
-    Err(TestError::NoFreePorts)
+    Ok(NodePorts { rpc_port: available_ports[0], p2p_port: available_ports[1] })
 }
 
 impl MadaraClient {
     async fn init(execution: ExecutionStrategy) -> Result<Self, TestError> {
-        let free_port = get_free_port()?;
+        let NodePorts { p2p_port, rpc_port } = find_available_ports()?;
 
         let manifest_path = Path::new(&env!("CARGO_MANIFEST_DIR"));
         let repository_root = manifest_path.parent().expect("Failed to get parent directory of CARGO_MANIFEST_DIR");
 
         std::env::set_current_dir(repository_root).expect("Failed to change working directory");
 
+        let madara_log = std::env::var("MADARA_LOG").unwrap_or_else(|_| "false".to_string());
+
         let child_handle = Command::new("cargo")
 		// Silence Madara stdout and stderr
 		.stdout(Stdio::null())
-		.stderr(Stdio::null())
+		.stderr(if madara_log == "true" { Stdio::inherit() } else { Stdio::null() })
 		.args([
 			"run",
 			"--release",
 			"--",
-			"--alice",
 			"--sealing=manual",
 			&format!("--execution={execution}"),
-			"--chain=dev",
-			"--tmp",
-			&format!("--rpc-port={free_port}")
+			"--dev",
+            "--tmp",
+			&format!("--port={p2p_port}"),
```

Comment:
why are we setting the p2p port? do we need reproducibility on those?

Reply:
Yes and also in case the default is not available